### PR TITLE
add Fuse support to FutureExt

### LINKED ISSUE #100: support fuse in FutureExt
Recently I switched from futures and futures-util to futures-lite. For the most case this was okay.

The only thing I did miss was the fuse functionality.

I implemented it here <https://github.com/plabayo/rama/blob/main/src/future.rs>.
Could I implement it in the actual futures-lite codebase.
Or what do you suggest?

### PR DESCRIPTION
**Pull request by GlenDC**
Closes #100

### CODE REVIEW

**Comment by GlenDC** [comment]
Ready for review @notgull

File: src/future.rs [hunk 1]
```diff
@@ -453,6 +453,36 @@ where
...
+
+impl<Fut: Future> Future for Fuse<Fut> {
+    type Output = Fut::Output;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Fut::Output> {
+        match self.as_mut().project().inner.as_pin_mut() {
+            Some(fut) => fut.poll(cx).map(|output| {
+                self.project().inner.set(None);
+                output
+            }),
```
Comment (notgull) [inline src/future.rs:480]: I'd prefer to move the `map`ping of `poll` to the expression being `match`ed on.

File: src/future.rs [hunk 2]
```diff
@@ -656,6 +686,29 @@ pub trait FutureExt: Future {
...
+    /// continue to do so for all future calls to `poll`).
+    ///
+    /// This combinator will drop the underlying future as soon as it has been
+    /// completed to ensure resources are reclaimed as soon as possible.
+    fn fuse(self) -> Fuse<Self>
+    where
+        Self: Sized,
+    {
+        Fuse::new(self)
+    }
```
Comment (notgull) [inline src/future.rs:710]: I would prefer a freestanding function instead of a member of the trait.

**Review by notgull — CHANGES_REQUESTED** [review]

File: src/future.rs (hunk 1 above)
Comment (GlenDC) [reply src/future.rs:480]: Can you elaborate? Not sure I follow.

Comment (GlenDC) [reply src/future.rs]: Ok @notgull I think I applied what you wanted here, but do let me know if that's still not what you were looking for here.

File: src/future.rs [hunk 3]
```diff
@@ -453,6 +453,62 @@ where
...
+        self.as_mut()
+            .project()
+            .inner
+            .as_pin_mut()
+            .map(|f| f.poll(cx))
+            .unwrap_or(Poll::Pending)
+            .map(|output| {
+                self.project().inner.set(None);
+                output
+            })
```
Comment (notgull) [inline src/future.rs:508]:
```suggestion
        match self.as_mut().project().inner.as_pin_mut().map(|f| f.poll(cx)) {
            Some(Poll::Ready(output)) => {
                self.project().inner.set(None);
                Poll::Ready(output)
            }
            
            Some(Poll::Pending) | None => Poll::Pending
        }
```

Apologies, I was on the train. I was moreso hoping for this

**Review by notgull — CHANGES_REQUESTED** [review]

File: src/future.rs (hunk 3 above)
Comment (GlenDC) [reply src/future.rs:508]: Cool. Looks good.

**Review by notgull — APPROVED** [review]
Thanks!

**Pull request merged**